marimo/_server/api/endpoints/terminal.py

```diff
@@ -48,15 +48,33 @@ async def _read_from_pty(master: int, websocket: WebSocket) -> None:
 
 
 async def _write_to_pty(master: int, websocket: WebSocket) -> None:
-    with os.fdopen(master, "wb", buffering=0) as master_file:
-        while True:
-            try:
-                data = await websocket.receive_text()
-                LOGGER.debug("Received: %s", data)
-                master_file.write(data.encode())
-                master_file.flush()
-            except (asyncio.CancelledError, WebSocketDisconnect):
-                break
+    try:
+        buffer = ""
+        with os.fdopen(master, "wb", buffering=0) as master_file:
+            while True:
+                try:
+                    data = await websocket.receive_text()
+                    LOGGER.debug("Received: %s", data)
+
+                    buffer += data
+                    if data in ["\r", "\n"]:  # Check for line ending
```

Comment:
Will fail if `exit;`, also there a few other ways a session can end; e.g. an EOT character 

I handled this by restarting in a `finally` with the process https://github.com/marimo-team/marimo/blob/71040b86daf649ce355a89fd0e422c62c8373a98/marimo/_server/api/endpoints/xterm.py#L188

Reply:
It actually still works, but it just prints a bunch of garbage as the terminal closes.

Reply:
there is a stack trace thrown that i am unsure how to catch, but it doesnt effect closing